marathon_acme/sse_protocol.py

```diff
@@ -0,0 +1,149 @@
+from twisted.internet import error
+from twisted.internet.protocol import Protocol
+
+
+class SseProtocol(Protocol):
+    """
+    A protocol for Server-Sent Events (SSE).
+    https://html.spec.whatwg.org/multipage/comms.html#server-sent-events
+    """
+
+    _buffer = b''
+    MAX_LENGTH = 16384
+
+    def __init__(self):
+        """
+        Initialize the protocol.
+
+        :param finished:
+            A deferred that will be called when the connection is closed
+        :param callbacks:
+            A dict mapping event types to functions that will handle their data
+        """
+        self.finished = None
+        self.callbacks = {}
+
+        self._reset_event_data()
+
+    def _reset_event_data(self):
+        self.event = 'message'
+        self.data = ''
+
+    def set_finished_deferred(self, d):
+        self.finished = d
+
+    def add_callback(self, event, callback):
+        self.callbacks[event] = callback
```

Comment:
Maybe this should be `set_callback`? `add_callback` implies that you can have more than one callback for a given event. (I don't have strong feelings about this, though.)